README.md
[![Build Status](https://travis-ci.org/ujh/iomrascalai.svg?branch=master)](https://travis-ci.org/ujh/iomrascalai)
[![Coverage Status](https://coveralls.io/repos/ujh/iomrascalai/badge.svg?branch=master)](https://coveralls.io/r/ujh/iomrascalai?branch=master)

[![Huboard](https://img.shields.io/github/issues/ujh/iomrascalai.svg?label=Issues%20%28HuBoard%29)](https://huboard.com/ujh/iomrascalai)
[![Gitter chat](https://badges.gitter.im/ujh/iomrascalai.png)](https://gitter.im/ujh/iomrascalai)


Iomrascálaí
===========

Iomrascálaí (Gaelic for wrestler, [see here for the pronunciation](https://raw.githubusercontent.com/ujh/iomrascalai/master/pronunciation.mp4)) is an AI for the game of Go/Weiqi/Baduk written in [Rust](https://www.rust-lang.org).

Installation
------------

If you just want to play a game against the engine you can download precompiled executables of the [releases](https://github.com/ujh/iomrascalai/releases) for Mac, Linux, and Windows. You will also need a graphical interface for playing games against the AI. We suggest downloading [GoGui](http://sourceforge.net/projects/gogui/).

Compilation
------------

Iomrascálaí requires the latest unstable (also called nightly) Rust compiler as well as the latest Cargo. Both are generally installed when you download the installers from the [Rust homepage](https://www.rust-lang.org).

Playing games
-------------

Once you've installed the nightly Rust compiler, Cargo, and [GoGui](http://gogui.sourceforge.net/) you can use the following shell scripts to compile the program and play games. All these scripts play a game on a 9x9 board with a time limit of 5 minutes by default. You can also play games on 13x13 and 19x19 by passing 13 or 19 as an argument to the scripts.

* `bin/play` will compile the program and start a game in [GoGui](http://gogui.sourceforge.net/). By default it will assign black to Iomrascálaí. The defaults can be changed easily by editing some constants in the script.
* `bin/play-gnugo` will compile the program and start a game against GnuGo. It will assign black to [GnuGo](https://www.gnu.org/software/gnugo/) and the game can be observed in [GoGui](http://gogui.sourceforge.net/). Again, the defaults can be changed by editing the script.
* `bin/play-self` will compile the program and start a game between two copies of Iomrascálaí. Just like with the other scripts the game can be observed in [GoGui](http://gogui.sourceforge.net/) and the parameters can be adjusted by editing the script.

Program parameters
------------------

Many parameters of Iomrascálaí can be changed and those changes directly affect the program strength.

You set the parameters by supplying a [TOML](https://github.com/toml-lang/toml) formatted configuration file when starting the program with either the `-c` or `--config` command line flag. A great way to get started is to capture the output of `-d` (or `--dump`) into a file and edit the variables. This is the default configuration and lists all possible variables with their default values. See the [api documentation](http://bettong.net/iomrascalai/) for details on what these variables do. Just use the search on top and enter the name of the configuration variable. It should take you to the page that lists the struct that defines it (e.g. [PriorsConfig](http://bettong.net/iomrascalai/api/iomrascalai/struct.PriorsConfig.html) for a variable in the `[priors]` block in the config file).

Development
===========

See the [issues](https://github.com/ujh/iomrascalai/issues) or [HuBoard](https://huboard.com/ujh/iomrascalai) for
planned features and bugs. [Join the chat](https://gitter.im/ujh/iomrascalai) for discussions and help.

Please note that this project is released with a Contributor Code of Conduct. By participating in this project you agree to abide by its terms.

License
=======

Iomrascálaí is licensed under GPLv3. See the
[LICENSE](https://github.com/ujh/iomrascalai/blob/master/LICENSE)
file for the complete license text. The scripts in `bin/` and `misc/clop` are licensed under the MIT license.


CHANGELOG.md
# master

At db420564342099369003b8f1e466e7e20bc4d4b1 the performace is:

```
refactorings-9x9.dat:
		83.8% wins (481 games of 574, ± 3.0 at 95%, ± 3.06 at 99%)
		73.14% same score as GnuGo (354 of 484, ± 3.93 at 95%, ± 4.01 at 99%)
refactorings-13x13.dat:
		42.0% wins (420 games of 1000, ± 3.04 at 95%, ± 3.11 at 99%)
		64.52% same score as GnuGo (280 of 434, ± 4.48 at 95%, ± 4.57 at 99%)
refactorings-15x15.dat:
		17.22% wins (104 games of 604, ± 3.0 at 95%, ± 3.06 at 99%)
		57.41% same score as GnuGo (62 of 108, ± 9.28 at 95%, ± 9.47 at 99%)
refactorings-17x17.dat:
		2.0% wins (2 games of 100, ± 2.73 at 95%, ± 2.79 at 99%)
		50.0% same score as GnuGo (1 of 2, ± 68.94 at 95%, ± 70.36 at 99%)
refactorings-19x19.dat:
		0.0% wins (0 games of 100, ± 0.0 at 95%, ± 0.0 at 99%)
		NaN% same score as GnuGo (0 of 0, ± NaN at 95%, ± NaN at 99%)
```

## 0.3.2 [☰](https://github.com/ujh/iomrascalai/compare/0.3.1...0.3.2)

* Fixes a bug where the bot passed too early and lost won games because of it.
  #261
* Resign lost games earlier. This speeds up the benchmark and is also nicer for
  humans. #264
* Speed up the calculations of the priors by moving them into the worker
  threads. #275
* Implement the `reg_genmove` GTP command. #289
* Implement the `kgs-genmove_cleanup` GTP command. #290
* Improve the scoring accuracy when the game is over. #292
* Fix bug where some rotations of the small patterns were missing. #299
* Implemented the `imrscl-uct_gfx` custom GTP command. A GoGui analyze
  compatible command to show the number of playouts that went through each child
  node of the root (i.e. each point on the board) #293

### Performance

*The benchmarks were run on a c4.8xlarge EC2 instance with 8 worker threads.*

After running 500 games on 9x9 with komi 6.5 and a time limit of 2
minutes (sudden death) the win rate against GnuGo 3.8 level 0 was
**76.2% (± 3.71 at 95%, ± 3.79 at 99%)** with the default
configuration. It agreed on the final score with GnuGo in **74.09% (± 4.35 at
95%, ± 4.44 at 99%)** of games.

After running 1000 games on 13x13 with komi 6.5 and a time limit of 10
minutes (sudden death) the win rate against GnuGo 3.8 level 0 was
**42.8% (± 3.05 at 95%, ± 3.11 at 99%)** with the default
configuration. It agreed on the final score with GnuGo in **59.78% (± 4.53 at
95%, ± 4.63 at 99%)** of games.


## 0.3.1 [☰](https://github.com/ujh/iomrascalai/compare/0.3.0...0.3.1)

* Implemented `gogui-analyze_commands` and `imrscl-safe_territory`
  which shows the intersections on the board that the bot thinks are
  safe (i.e. definitely belong to one color).
* Stop playing when all intersections are decided (i.e. the bot thinks
  they belong to a certain color).
* Implement `final_status_list`
* Implement Chinese scoring
* Return to command line arguments for turning on/off logging, GoGui
  effects, and ruleset configuration.

### Performance

*Note that we now use Chinese rules instead of CGOS/Tromp Taylor rules
 which may change the results.*

After running 500 games on 9x9 with komi 6.5 and a time limit of 5
minutes (sudden death) the win rate against GnuGo 3.8 level 0 was
**77.4% (± 3.65 at 95%, ± 3.72 at 99%)** with the default
configuration.

After running 500 games on 13x13 with komi 6.5 and a time limit of 10
minutes (sudden death) the win rate against GnuGo 3.8 level 0 was
**35.2% (± 4.16 at 95%, ± 4.25 at 99%)** with the default
configuration.


## 0.3.0 [☰](https://github.com/ujh/iomrascalai/compare/0.2.4...0.3.0)

* Added 3x3 pattern support. These are used as priors in the tree and
  during the playouts.
* Added the RAVE heuristic to the search.
* Terminate search early if the best move is ahead by a lot.
* Changed the way to change the configuration of the engine.
  Everything is now stored inside a
  [TOML](https://github.com/toml-lang/toml) file instead of having
  command line switches for everything.
* Measure the playouts per second per thread and display it when
  logging is turned on.
* Scripts to run [CLOP](http://www.remi-coulom.fr/CLOP/) parameter
  optimization.
* Removed all weak engines (random, AMAF, MC).
* Clarified the licensing (scripts are MIT licensed).
* Added a code of conduct.

### Performance

*Note that the measurement of the error margins have changed. We now
 calculate proper 95% and 99% confidence intervals whereas previously
 we only used the ~68% standard deviation.*

After running 500 games on 9x9 with komi 6.5 and a time limit of 5
minutes (sudden death) the win rate against GnuGo 3.8 level 0 was
**85.40% (± 3.08 at 95%, ± 3.14 at 99%)** with the default
configuration.

After running 500 games on 13x13 with komi 6.5 and a time limit of 10
minutes (sudden death) the win rate against GnuGo 3.8 level 0 was
**36.6% (± 4.2 at 95%, ± 4.29 at 99%)** with the default
configuration.


## 0.2.4 [☰](https://github.com/ujh/iomrascalai/compare/0.2.3...0.2.4)

This release doesn't contain any new features. But as the last release
happened quite a long time ago and there were some (small) internal
changes a new release seems to be in order. Also the Rust compiler
changed a lot in between and changed the performance.

### Performance

After running 200 games on 9x9 with komi 6.5 and a time limit of 5
minutes (sudden death) the win rate against GnuGo 3.8 level 0 was
**65.5% ± 3.4** for the default engine with 8 threads.

After running 200 games on 13x13 with komi 6.5 and a time limit of 10
minutes (sudden death) the win rate against GnuGo 3.8 level 0 was
**10% ± 2.1** for the default engine with 8 threads.

## 0.2.3 [☰](https://github.com/ujh/iomrascalai/compare/0.2.2...0.2.3)

* Playout the aftermath by default on CGOS
* Play in the middle of big eyes during playouts
* Resign lost games faster (to save time)

### Performance

After running 200 games on 9x9 with komi 6.5 and a time limit of 5
minutes (sudden death) the win rate against GnuGo 3.8 level 0 was **75.5%
± 3%** for the default engine with 8 threads.


src/gtp/mod.rs
/************************************************************************
 *                                                                      *
 * Copyright 2014 Thomas Poinsot, Urban Hafner                          *
 * Copyright 2015 Urban Hafner, Thomas Poinsot, Igor Polyakov           *
 * Copyright 2016 Urban Hafner                                          *
 *                                                                      *
 * This file is part of Iomrascálaí.                                    *
 *                                                                      *
 * Iomrascálaí is free software: you can redistribute it and/or modify  *
 * it under the terms of the GNU General Public License as published by *
 * the Free Software Foundation, either version 3 of the License, or    *
 * (at your option) any later version.                                  *
 *                                                                      *
 * Iomrascálaí is distributed in the hope that it will be useful,       *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
 * GNU General Public License for more details.                         *
 *                                                                      *
 * You should have received a copy of the GNU General Public License    *
 * along with Iomrascálaí.  If not, see <http://www.gnu.org/licenses/>. *
 *                                                                      *
 ************************************************************************/

use board::Color;
use board::Move;
use config::Config;
use engine::Engine;
use engine::EngineController;
use game::Game;
use ruleset::Ruleset;
use sgf::parser::Parser;
use timer::Timer;
use version;

use regex::Regex;
use std::path::Path;
use std::sync::Arc;
use time::precise_time_ns;

pub mod driver;
mod test;

pub struct GTPInterpreter<'a> {
    byo_stones: i32,
    byo_time: i64,
    commands: Vec<&'a str>,
    config: Arc<Config>,
    controller: EngineController,
    game: Game,
    main_time: i64,
    running: bool,
    timer: Timer,
}

impl<'a> GTPInterpreter<'a> {
    pub fn new(config: Arc<Config>, engine: Engine) -> GTPInterpreter<'a> {
        let controller = EngineController::new(config.clone(), engine);
        let komi = 6.5;
        let boardsize = 19;
        let commands = vec![
            "boardsize",
            "clear_board",
            "final_score",
            "final_status_list",
            "genmove",
            "gogui-analyze_commands",
            "imrscl-donplayouts",
            "imrscl-ownership",
            "imrscl-uct_gfx",
            "kgs-genmove_cleanup",
            "known_command",
            "komi",
            "list_commands",
            "loadsgf",
            "name",
            "play",
            "protocol_version",
            "quit",
            "reg_genmove",
            "showboard",
            "time_left",
            "time_settings",
            "version",
            ];
        GTPInterpreter {
            byo_stones: 0,
            byo_time: 0,
            commands: commands,
            config: config.clone(),
            controller: controller,
            game: Game::new(boardsize, komi, config.ruleset),
            main_time: 5,
            running: true,
            timer: Timer::new(config),
        }
    }

    pub fn quit(&mut self) {
        self.running = false;
    }

    pub fn komi(&self) -> f32 {
        self.game.komi()
    }

    pub fn ruleset(&self) -> Ruleset {
        self.config.ruleset
    }

    pub fn boardsize(&self) -> u8 {
        self.game.size()
    }

    pub fn read(&mut self, input: &str) -> Result<String, String> {
        let preprocessed = self.preprocess(input);
        if preprocessed.len() == 0 {
            return Err("empty command".to_string())
        };
        let command: Vec<&str> = preprocessed.split(' ').collect();
        if !self.commands.contains(&command[0]) {
            return Err("unknown command".to_string());
        }
        let arguments = &command[1..];
        match command[0] {
            "boardsize" => self.execute_boardsize(arguments),
            "clear_board" => self.execute_clear_board(arguments),
            "final_score" => self.execute_final_score(arguments),
            "final_status_list" => self.execute_final_status_list(arguments),
            "genmove" => self.execute_genmove(arguments),
            "gogui-analyze_commands" => self.execute_gogui_analyze_commands(arguments),
            "imrscl-donplayouts" => self.execute_imrscl_donplayouts(arguments),
            "imrscl-ownership" => self.execute_imrscl_ownership(arguments),
            "imrscl-uct_gfx" => self.execute_uct_gfx(arguments),
            "kgs-genmove_cleanup" => self.execute_kgs_genmove_cleanup(arguments),
            "known_command" => self.execute_known_command(arguments),
            "komi" => self.execute_komi(arguments),
            "list_commands" => self.execute_list_commands(arguments),
            "loadsgf" => self.execute_loadsgf(arguments),
            "name" => self.execute_name(arguments),
            "play" => self.execute_play(arguments),
            "protocol_version" => self.execute_protocol_version(arguments),
            "quit" => self.execute_quit(arguments),
            "reg_genmove" => self.execute_reg_genmove(arguments),
            "showboard" => self.execute_showboard(arguments),
            "time_left" => self.execute_time_left(arguments),
            "time_settings" => self.execute_time_settings(arguments),
            "version" => self.execute_version(arguments),
            _ => Err("unknown command".to_string())
        }

    }

    fn execute_name(&mut self, _: &[&str]) -> Result<String, String> {
        Ok("Iomrascalai".to_string())
    }

    fn execute_version(&mut self, _: &[&str]) -> Result<String, String> {
        Ok(version::version().to_string())
    }

    fn execute_protocol_version(&mut self, _: &[&str]) -> Result<String, String> {
        Ok("2".to_string())
    }

    fn execute_list_commands(&mut self, _: &[&str]) -> Result<String, String> {
        Ok(self.commands[1..].iter().fold(self.commands[0].to_string(), |acc, &el| format!("{}\n{}", acc, el)))
    }

    fn execute_known_command(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(0) {
            Some(comm) => Ok(format!("{}", self.commands.contains(comm))),
            None => Err("missing argument".to_string())
        }
    }

    fn execute_boardsize(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(0) {
            Some(comm) => match comm.parse::<u8>() {
                Ok(size) => {
                    self.game = Game::new(size, self.komi(), self.ruleset());
                    Ok("".to_string())
                },
                Err(e) => Err(format!("{:?}", e))
            },
            None => Err("missing argument".to_string())
        }
    }

    fn execute_clear_board(&mut self, _: &[&str]) -> Result<String, String> {
        let size = self.boardsize();
        let komi = self.komi();
        self.game = Game::new(size, komi, self.ruleset());
        self.timer.setup(self.main_time, self.byo_time, self.byo_stones);
        self.controller.reset(size, komi);
        Ok("".to_string())
    }

    fn execute_komi(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(0) {
            Some(comm) =>
                match comm.parse::<f32>() {
                    Ok(komi) => {
                        self.game.set_komi(komi);
                        Ok("".to_string())
                    },
                    Err(e) => Err(format!("{:?}", e))
                },
            None => Err("missing argument".to_string())
        }
    }

    fn execute_reg_genmove(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(0) {
            Some(c) => {
                let started_at = precise_time_ns();
                self.timer.start(&self.game);
                let color = Color::from_gtp(c);
                let (m, playouts) = self.controller.genmove(color, &self.game, &self.timer);
                let response = match self.game.play(m) {
                    Ok(_) => {
                        self.timer.reset();
                        Ok(m.to_gtp())
                    },
                    Err(e) => {
                        Err(format!("Illegal move {:?} ({:?})", m, e))
                    }
                };
                Self::measure_playout_speed(started_at, playouts, &self.config);
                response
            },
            None => Err("missing argument".to_string())
        }
    }

    fn execute_genmove(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(0) {
            Some(comm) => {
                let started_at = precise_time_ns();
                self.timer.start(&self.game);
                let color = Color::from_gtp(comm);
                let (m, playouts) = self.controller.genmove(color, &self.game, &self.timer);
                let response = match self.game.play(m) {
                    Ok(g) => {
                        self.game = g;
                        self.timer.stop();
                        Ok(m.to_gtp())
                    },
                    Err(e) => {
                        Err(format!("Illegal move {:?} ({:?})", m, e))
                    }
                };
                Self::measure_playout_speed(started_at, playouts, &self.config);
                response
            },
            None => Err("missing argument".to_string())
        }
    }

    fn execute_kgs_genmove_cleanup(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(0) {
            Some(comm) => {
                self.game.reset_game_over();
                let started_at = precise_time_ns();
                self.timer.start(&self.game);
                let color = Color::from_gtp(comm);
                let (m, playouts) = self.controller.genmove_cleanup(color, &self.game, &self.timer);
                let response = match self.game.play(m) {
                    Ok(g) => {
                        self.game = g;
                        self.timer.stop();
                        Ok(m.to_gtp())
                    },
                    Err(e) => {
                        Err(format!("Illegal move {:?} ({:?})", m, e))
                    }
                };
                Self::measure_playout_speed(started_at, playouts, &self.config);
                response
            },
            None => Err("missing argument".to_string())
    	}
    }

    fn execute_imrscl_donplayouts(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(0) {
            Some(playouts_str) => {
                match playouts_str.parse() {
                    Ok(playouts) => {
                        self.controller.donplayouts(&self.game, playouts);
                        Ok("".to_string())
                    },
                    Err(e) => Err(format!("{:?}", e))
                }
            }
            None => Err("missing argument".to_string()),
        }
    }

    fn execute_imrscl_ownership(&mut self, _: &[&str]) -> Result<String, String> {
        let stats = self.controller.ownership_statistics();
        Ok(stats)
    }

    fn execute_play(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(1) {
            Some(second) => {
                let m = Move::from_gtp(arguments[0], second);
                match self.game.play(m) {
                    Ok(g) => {
                        self.game = g;
                        Ok("".to_string())
                    },
                    Err(e) => Err(format!("Illegal move {:?} ({:?})", m, e))
                }
            },
            None => Err("missing argument".to_string())
        }
    }

    fn execute_showboard(&mut self, _: &[&str]) -> Result<String, String> {
        Ok(format!("\n{}", self.game))
    }

    fn execute_quit(&mut self, _: &[&str]) -> Result<String, String> {
        self.quit();
        Ok("".to_string())
    }

    fn execute_final_score(&mut self, _: &[&str]) -> Result<String, String> {
        self.game.reset_game_over();
        Ok(self.controller.final_score(&self.game))
    }

    fn execute_final_status_list(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(0) {
            Some(kind) => {
                self.game.reset_game_over();
                self.controller.final_status_list(&self.game, kind)
            },
            None => Err("missing argument".to_string())
        }
    }

    fn execute_time_settings(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(2) {
            Some(third) => {
            	//command[1] and command[2] should be there
                match (arguments[0].parse::<i64>(), arguments[1].parse::<i64>(), third.parse::<i32>()) {
                    (Ok(main), Ok(byo), Ok(stones)) => {
                        self.main_time = main;
                        self.byo_time = byo;
                        self.byo_stones = stones;
                        self.timer.setup(main, byo, stones);
                        Ok("".to_string())
                    }
                    _ => Err("error parsing time_settings".to_string())
                }
            },
            None => Err("missing argument(s)".to_string())
        }
    }

    fn execute_uct_gfx(&mut self, _: &[&str]) -> Result<String, String> {
        let stats = self.controller.uct_gfx();
        Ok(stats)
    }

    fn execute_time_left(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(2) {
            Some(third) => {
                match (arguments[1].parse::<i64>(), third.parse::<i32>()) {
                    (Ok(time), Ok(stones)) => {
                        self.timer.update(time, stones);
                        Ok("".to_string())
                    },
                    _ => Err("error parsing time_left".to_string())
                }
            },
            None => Err("missing argument(s)".to_string())
        }
    }

    fn execute_loadsgf(&mut self, arguments: &[&str]) -> Result<String, String> {
        match arguments.get(0) {
            Some(comm) => {
                let filename = comm;

                match Parser::from_path(Path::new(filename)) {
                    Ok(parser) => {
                    	let game = parser.game();
                        match game {
                            Ok(g) => {
                                self.game = g;
                                Ok("".to_string())
                            },
                            Err(_) => Err("cannot load file".to_string())
                        }
                    },
                    Err(_) => Err("cannot load file".to_string())
                }
            },
            None => Err("missing argument".to_string())
        }
    }

    fn execute_gogui_analyze_commands(&mut self, _: &[&str]) -> Result<String, String> {
        let analyze_commands = vec![
            "dboard/Ownership/imrscl-ownership",
            "plist/Final Status List Dead/final_status_list dead",
            "plist/Final Status List Alive/final_status_list alive",
            "gfx/Uct Gfx/imrscl-uct_gfx"
                ];
        Ok(analyze_commands[1..].iter().fold(analyze_commands[0].to_string(), |acc, &el| format!("{}\n{}", acc, el)))
    }

    fn measure_playout_speed(started_at: u64, playouts: usize, config: &Arc<Config>) {
        let finished_at = precise_time_ns();
        let duration_ns = finished_at - started_at;
        let duration_s = (duration_ns as f64) / 1000000000.0;
        let pps = (playouts as f64) / duration_s;
        let threads = config.threads;
        let ptps = pps / (threads as f64);
        config.log(format!("{}pps ({}pps per thread)", pps.round() as usize, ptps.round() as usize));
    }

    fn preprocess(&self, input: &str) -> String {
        // Convert tab to space
        let horizontal_tab = Regex::new(r"\t").unwrap();
        let without_tabs = horizontal_tab.replace_all(input, " ");
        // Remove all control characters
        let cntrls = Regex::new(r"[:cntrl:]").unwrap();
        let without_ctrls = cntrls.replace_all(without_tabs.as_ref(), "");
        // Then we remove anything after a #
        let comment = Regex::new(r"#.*").unwrap();
        let without_comment = comment.replace(without_ctrls.as_ref(), "");
        // We remove the whitespaces before/after the string
        without_comment.trim().to_string()
    }
}

impl<'a> Drop for GTPInterpreter<'a> {

    fn drop(&mut self) {
        self.quit();
    }
}
